extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleAuthorization.scala

```diff
@@ -34,23 +35,22 @@ case class RuleAuthorization(spark: SparkSession) extends Authorization(spark) {
     val auditHandler = new SparkRangerAuditHandler
     val ugi = getAuthzUgi(spark.sparkContext)
     val (inputs, outputs, opType) = PrivilegesBuilder.build(plan, spark)
-    val requests = new ArrayBuffer[AccessRequest]()
+    var requests = new HashMap[(AccessResource, AccessType), AccessRequest]()
```

Comment:
Why `immutable.HashMap` instead of `mutable.HashMap` is used here?

BTW, Scala 2.13 changed the alias of `HashMap` from `mutable.HashMap` to `immutable.HashMap`, to make the code clear, please write `new immutable.HashMap` or use `new mutable.HashMap` inline

Reply:
and please leave a few comments here to explain the background

Reply:
and, was `immutable.Map.newBuilder` considered?

Reply:
1. Change to `mutable.HashMap`
2. Added some comments
3. I'm not sure if `immutable.Map.newBuilder` is better than `mutable.HashMap`?

Reply:
How about using `mutable.LinkedHashSet[AccessRequest]` as the type for `requests`?
1. it maintains the original adding order
2. it guarantees the element's uniqueness reliing on the `hashCode` 
3. good for element looping as fast as the array

And make sure related methods of AccessRequest properly implemented. Add some unit test for assuring it.

Reply:
I'm not sure if there is any overhead in maintaining the linked elements? And would the `mutable.LinkedHashSet[AccessRequest]` be slower than `ArrayBuffer` when looping the elements?

Reply:
The uniqueness is guaranteed by `hashCode`, which could be told from its name `HashSet` of `LinkedHashSet`. Not much overhead cost for maintaining the order inside, just adding a reference to a linked list. And the linked list perform well in looping elements.

Reply:
I mean `mutable.LinkedHashSet[AccessRequest]` may be slower than HashSet + ArrayBuffer in our case.

Reply:
the unique key is NOT the `AccessRequest` itself but part of properties, the combination of ArrayBuffer with a HashSet looks fine to me.

Reply:
> I mean `mutable.LinkedHashSet[AccessRequest]` may be slower than HashSet + ArrayBuffer in our case.

No, it wont be slowed down, as LinkedHashSet maintaining both HashSet and Linked list inside.

Reply:
> the unique key is NOT the `AccessRequest` itself but part of properties, the combination of ArrayBuffer with a HashSet looks fine to me.

Noticed. But the `HashSet` still reily on the same logic with `hashCode` inside. And no proper customed method of  either `AccessRequest` or its parent class.